test/Test/BaseDAO.hs

```diff
@@ -0,0 +1,166 @@
+
+module Test.BaseDAO
+  ( test_BaseDAO
+  ) where
+
+import Universum
+
+import qualified Data.Map.Strict as Map
+import Lorentz
+import Lorentz.Test
+import Morley.Nettest
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.HUnit (testCase)
+import Util.Named
+
+import qualified Lorentz.Contracts.Spec.FA2Interface as FA2
+import qualified Indigo.Contracts.SampleFA2 as SampleFA2 -- TODO: Replace SampleFA2 with BaseDAO
+
+-- TODO: Replace SampleFA2 with BaseDAO
+baseDAOStorage
+  :: Map (Address, FA2.TokenId) Natural
+  -> SampleFA2.Operators -> SampleFA2.Storage
+baseDAOStorage bal operators =
+  SampleFA2.mkStorage daoToken bal operators
+  where
+    daoToken :: [ FA2.TokenMetadata ]
+    daoToken =
+      [ FA2.TokenMetadata
+          { token_id = 0
+          , symbol = [mt|unfrozen_token|]
+          , name = [mt|Unfrozen Token|]
+          , decimals = 8
+          , extras = Map.empty
+          }
+      , FA2.TokenMetadata
+          { token_id = 1
+          , symbol = [mt|frozen_token|]
+          , name = [mt|Frozen Token|]
+          , decimals = 8
+          , extras = Map.empty
+          }
+      ]
+
+-- TODO: Replace SampleFA2 with BaseDAO
+baseDAOContract :: FA2.PermissionsDescriptor -> Contract FA2.Parameter SampleFA2.Storage
+baseDAOContract p =
+  SampleFA2.fa2Contract p
+
+originateBaseDAO
+  :: MonadNettest caps base m => FA2.PermissionsDescriptor
+  -> m ((Address, Address), (Address, Address), TAddress FA2.Parameter)
+originateBaseDAO p = do
+  owner1  :: Address <- newAddress "owner1"
+  operator1  :: Address <- newAddress "operator1"
+  owner2  :: Address <- newAddress "owner2"
+  operator2  :: Address <- newAddress "operator2"
+
+  let bal = Map.fromList
+        [ ((owner1, 0), 100)
+        , ((owner2, 0), 100)
+        ]
+  let (operators :: SampleFA2.Operators) =
+        BigMap $ Map.fromList
+          [ ((owner1, operator1), ())
+          , ((owner2, operator2), ())
+          ]
+  let
+    originateData = OriginateData
+      { odFrom = AddressResolved genesisAddress
+      , odName = "BaseDAO"
+      , odBalance = toMutez 0
+      , odStorage = baseDAOStorage bal operators
+      , odContract = baseDAOContract p
+      }
+  dao <- originate originateData
+  pure ((owner1, operator1), (owner2, operator2), dao)
+
+test_BaseDAO :: TestTree
+test_BaseDAO = testGroup "tests to check BaseDAO contract functionality"
+  [ testCase "Allows zero transfer from non-existent accounts" $
+      nettestTestExpectation zeroTransferScenario
+  , testCase "Is denied if only owner transfer is allowed" $
+      nettestTestExpectation ownerTransferAllowScenario
+  , testCase "aborts if there is a failure (due to low balance)" $
+      nettestTestExpectation lowBalanceScenario
+  -- TODO: testcase: "Allows zero transfer from non-existent accounts"
+  -- TODO: testcase: "Is denied if only owner transfer is allowed"
+  -- TODO: testcase: "aborts if there is a failure (due to low balance)"
```

Comment:
@gromakovsky I've refactored the tests as above.

It is not finished yet, my plan is to implement all the TODO test cases. 

Currently, I use `FA2Interface` and `SampleFA2` from morley-ledger with slight modification to help running the tests.

I would like your feedback whether or not this tests structure is good, and whether or not the TODO test cases cover all our needs or we want like to add more (or remove some).

Reply:
The structure looks good. I have some comments about these tests. Note that in this contract we have two token_ids, so in `balance_of` tests we should test both of them. Also there is this special behavior that should be tested:
> The administrator can transfer tokens from any address to any address.

Reply:
I've removed all the non-default permission descriptor test cases, add the admin transfer test case, as well as includes both tokens for `balance_of` tests. 

Hope to hear your feedback, Ivan.

Related to the CI, it fails due to this error: https://buildkite.com/serokell/basedao/builds/18#ccfe2eb7-9516-4e32-b535-bab5d59726d9, I'm not sure why.

Reply:
`indigo-0.2.2` is not present in `stackage.nix`/`hackage.nix` snapshot that we are using. I have updated them, the error should be fixed.